### PR TITLE
Introduce launch-args modification API

### PR DESCRIPTION
**Pull request by d4vidi**
## Description
A rewrite of #2618, #2613, introducing app launch-args editing according to the following syntax:
```js
device.appLaunchArgs.modify({
  newArg: 'aValue',
  existingArgToOverride: 'newValue',  
  existingArgToDelete: undefined,
});
```

and there's also:
```js
device.appLaunchArgs.reset();
device.appLaunchArgs.get();
```

In the future, we expect to support multiple apps according to the following terms:
- In a multi-apps environment, launch-args will be managed per each, separately (following the equivalent logic of the global Detox config).
- Arguments will be available via the same API, in the form of `device.appLaunchArg('app-name')`.
- In case `app-name` isn't explicitly specified -- will modify the currently "selected" app.

### CODE REVIEW

File: detox/index.d.ts [hunk 1]
```diff
@@ -274,6 +274,55 @@ declare global {
...
+             *
+             * @param modifier The modifier object.
+             *
+             * @example
+             * // With current launch arguments set to:
+             * // {
+             * //   mockServerPort: 1234,
+             * //   mockServerCredentials: 'user@test.com:12345678',
+             * // }
+             * device.modify({
```
Comment (noomorph) [inline detox/index.d.ts:299]: Not `device`, but `device.appLaunchArgs`

File: detox/index.d.ts [hunk 2]
```diff
@@ -274,6 +274,55 @@ declare global {
...
+             *   mockServerToken: 'abcdef',
+             * };
+             * await device.launchApp();
+             * // => launch-arguments become:
+             * // {
+             * //   mockServerPort: 4321,
+             * //   mockServerToken: 'abcdef',
+             * // }
+             */
+            modify(modifier: object): void;
```
Comment (noomorph) [inline detox/index.d.ts:311]: Maybe `assign`, to be in line with `Object.assign` ?

File: detox/index.d.ts (hunk 1 above)
Comment (d4vidi) [reply detox/index.d.ts:299]: 💪🏻

File: detox/index.d.ts (hunk 2 above)
Comment (d4vidi) [reply detox/index.d.ts:311]: Well my line of though is that since for the undefined/null case we effectively delete properties, assigning isn't 100% descriptive. We're effectively modifying, perhaps even editing 🤔

File: detox/index.d.ts [hunk 3]
```diff
@@ -785,8 +866,8 @@ declare global {
              */
             delete?: boolean;
             /**
-             * Detox can start the app with additional launch arguments
-             * The added launchArgs will be passed through the launch command to the device and be accessible via [[NSProcessInfo processInfo] arguments]
+             * Arguments to pass-through into the app.
+             * Refer to the [dedicated guide](https://github.com/wix/Detox/blob/master/docs/APIRef.LaunchArgs.md) for complete details.
              */
             launchArgs?: any;
```
Comment (noomorph) [inline detox/index.d.ts:872]: @d4vidi, don't you find this weird? I don't remember we ever had such an API... 😮 🤔

File: detox/src/devices/Device.js [hunk 4]
```diff
@@ -49,7 +50,7 @@ class Device {
     }
 
     const baseLaunchArgs = {
-      ...this._deviceConfig.launchArgs,
+      ...this._launchArgs.get(),
```
Comment (noomorph) [inline detox/src/devices/Device.js:53]: @d4vidi , I am confused — so how I am supposed to suppress base launchArgs if passing `null` or `undefined` simply removes properties inside `this._launchArgs`?

File: detox/src/devices/LaunchArgs.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+  modify(launchArgs) {
+    Object
+      .keys(launchArgs)
+      .forEach((name) =>
+        this._setLaunchArg(name, launchArgs[name]));
+  }
+
+  reset() {
+    this._args = {};
```
Comment (noomorph) [inline detox/src/devices/LaunchArgs.js:16]: Wait, wait — are we resetting to an empty object or to the original `launchArgs`? This looks like a discrepancy.

File: detox/src/devices/Device.js (hunk 4 above)
Comment (noomorph) [reply detox/src/devices/Device.js:53]: Why do we need an extra complexity here? Can't we just have it as a blank key-value map and merge at the very last stage all these three: `this._deviceConfig.launchArgs, this._launchArgs.get(), ...params.launchArgs`?

Is the goal to make sure users can dynamically read the launch args? Do we really need this much flexibility?

Comment (noomorph) [reply detox/src/devices/Device.js:53]: Okay, now I see. You want to be able to reset all launchArgs forcibly. Makes sense.

File: detox/test/detox.config.js [hunk 6]
```diff
@@ -99,7 +99,12 @@ module.exports = {
       type: "android.emulator",
       device: {
         avdName: "Pixel_API_28"
-      }
+      },
+      launchArgs: {
```
Comment (noomorph) [inline detox/test/detox.config.js:103]: This is JavaScript, so I'd advise reusing a single `launchArgs` object across all the configurations.

File: detox/test/e2e/22.launch-args.test.js [hunk 7]
```diff
@@ -26,25 +26,55 @@ describe(':android: Launch arguments', () => {
...
     await device.launchApp({newInstance: true, launchArgs});
-
     await element(by.text('Launch Args')).tap();
 
     await assertLaunchArg('complex', JSON.stringify(launchArgs.complex));
     await assertLaunchArg('complexlist', JSON.stringify(launchArgs.complexlist));
   });
 
+  it('should allow for arguments modification', async () => {
+    device.appLaunchArgs.modify({
```
Comment (noomorph) [inline detox/test/e2e/22.launch-args.test.js:48]: Could we add an assertion regarding the original appLaunchArgs values? ... i.e.:

```js
if (!_.isEqual(device.appLaunchArgs.get(), { app: 'le', goo: 'gle', micro: 'soft' }) {
  fail('Please check detox.config.js - your configuration should have custom app launchArgs');
}
```

File: docs/APIRef.LaunchArgs.md [hunk 8]
```diff
@@ -0,0 +1,43 @@
+# Launch Arguments
+
+Detox allows in its [`device` API](APIRef.DeviceObjectAPI.md) for the launching the app under test in the test device via an explicit call to `device.launchApp()`. Through various means, Detox enables specifying a set of user-defined arguments (key-value pairs) to be passed on to the app when launched, so as to make them available inside the app itself at runtime (both on the native side, and - if applicable, on the Javascript side).
```
Comment (noomorph) [inline docs/APIRef.LaunchArgs.md:3]: Great effort, really, but should I mention that the article is a bit too wordy?
We can revisit it later if we have the capacity, though.

Comment (noomorph) [reply docs/APIRef.LaunchArgs.md:3]: Nevertheless, kudos for writing this! 👍

File: detox/index.d.ts (hunk 2 above)
Comment (noomorph) [reply detox/index.d.ts:311]: Okay.

File: detox/index.d.ts (hunk 3 above)
Comment (d4vidi) [reply detox/index.d.ts:878]: hmmm wdym?

Comment (noomorph) [reply detox/index.d.ts:878]: Check the actual implementation of `Device.js` and find it.

Comment (noomorph) [reply detox/index.d.ts:878]: Oh, Github you're drunk. :( That was just 306 / 362 line jump. No worries, resolving.

**Review by noomorph — APPROVED** [review]
